manic/repository_git.py

```diff
@@ -103,7 +103,25 @@ def _clone_repo(self, base_dir_path, repo_dir_name, verbosity):
         """
         cwd = os.getcwd()
         os.chdir(base_dir_path)
-        self._git_clone(self._url, repo_dir_name, verbosity)
+
+        # get url
+        url = self._url
+
+        # add port
+        if self._port is not EMPTY_STR:
+            ind1 = url.find('//')+2
+            ind2 = url[ind1:].find('/')
+            url = url[:ind1+ind2] + ':' + self._port + url[ind1+ind2:]
+
+        # add user name
+        if self._user_name is not EMPTY_STR:
+            ind1 = url.find('//')+2
+            url = url[:ind1] + self._user_name + '@' + url[ind1:]
+
+        # update url
+        self._url = url
```

Comment:
It feels wrong to me to update self._url here: It feels like self._url should be fixed after initialization, otherwise you have the confusing situation that self._url changes over time, depending on whether `self._clone_repo` happens to have been called yet.

I'd suggest introducing a new function on the repository class that builds this full url, including port and user name. Then you can either call it from the `__init__` method, setting `self._url` there, or else you can call it from wherever `self._url` is currently accessed (replacing direct references to `self._url` with calls to this new function, like `self._get_resolved_url()`.

This begs a couple of questions:

(1) Should this fully resolved url be used everywhere where `self._url` currently appears in repository_git.py, or are there some uses that should continue to use the original, unresolved url?

(2) What about repository_svn? Should this functionality be implemented there as well? If not, then please add some error checking that notes that user_name and port cannot be specified for svn repositories.